### DISCUSSION TITLE
Proposal: Include CSS-Variables in theme

### DISCUSSION THREAD

**Discussion by scottyeck** [mantinedev/mantine#3288] [Feature requests]
Hey Mantine team - thanks for all your hard work. Great stuff.

### Context

When debugging a design implementation using a CSS-in-JS framework (such as Mantine), translating between a _raw value_ shown in Dev Tools and a _system value_ specified in a Design spec can be challenging.

<details>
<summary>Example scenario</summary>

### Example scenario

We can reference STYLE ATTRIBUTES in two ways:

* Their _raw value_: (i.e. for colors, a hex or RGB string)
* Their _system value_: (i.e. for colors, `${colorName}.{rank}`, `blue.4`)

Consider this flow...

* Step 1: Get raw value in use
  * Observe (possibly) wrong value (`color`, `font-size`, etc.) in web UI
  * Open up developer tools to identify raw value via CSS inspector.
* Step 2: Get prescribed system value
  * Open up design spec doc (Figma, Sketch, etc.) and find corresponding component.
  * Identify specified system value
* Step 3: Compare in-use raw value with prescribed system value.
  * Open up design system docs
  * Find mapped _raw value_ for specified _system value_ and compare.

Does this work? Sure. Is it a pain? Definitely.

</details>

### Problem

👉 **Mantine could use CSS Variables to simplify the debugging UX but currently does not.**

Currently when the `withCSSVariables` prop is specified on the `<MantineProvider>`, Mantine merely builds up a mapping of CSS variable names to theme values and injects an `emotion` `<Global>` style in which those variables are then declared. The `theme` itself is left raw, so CSS developer tools reference raw values rather than system values.

### Proposed Solution

👉 **Instead of simply building up the CSS variables mapping, Mantine _could_ then transform the theme itself to reference the CSS variables rather than the raw values.** This way, at render-time, emotion would specify a CSS variable (representing a system value) visible in developer tooling for a high-quality debugging experience. :sunglasses:

Implementation-wise, we'd probably want to separate this out from `withCSSVariables` interface since the prescription of "include CSS variables in my project" varies drastically from "mutate my entire theme" and it could be dangerous (see question below).

### Questions

_(To start obvious ones - is this a real problem? Has it been discussed already? Is there prior art / relevant discussion? Is the proposed solution sound?)_

Actual meaty questions...

* I haven't dug on this yet, but I'd guess there may be scenarios where mantine is use a raw value on the theme to derive a new value at the component-level. (Darkening / lightening colors are an obvious one here.)
  * Is this guess accurate?
  * To get around it, we'd probably have to maintain two instances of the theme - one with the variable names and one with the raw values.

* There's possible performance considerations when we ask the CSSOM to compute variable values rather than specifying raw values. I'm entirely uninformed here, though if the API allows for this to be opt-in only, should be fine.

**Comment by scottyeck** [comment]
Should also note I'm more than happy to work on this - just would love a blessing before I get started. 😇

**Comment by rtivital** [comment]
I do not really see much value in CSS variables – you can debug components by referencing source code. Implementing this would require a lot of breaking changes, which is not something that I want.
